### PR TITLE
fix(ttnn): handle zero inputs in prod_bw gradient calculation (#54551)

### LINKED ISSUE #54551: [Bounty $1,000] ttnn.prod_bw returns non-finite gradients for zero inputs
## Summary

`ttnn.prod_bw` computes the gradient as

```text
prod(input) * grad / input
```

which is `reciprocal(input)` multiplied by a broadcast `prod(input) * grad`. Where the input
contains an exact zero, `reciprocal(0)` is `+inf` and the surviving product is `0`, so the
gradient at that position is non-finite instead of the finite value autograd defines.

```text
input:    [2, 0, 4]
grad:     1
expected: [0, 8, 0]
current:  [0, non-finite, 0]
```

This is a source-level analysis: I do not have a card, so the host replay below renders the
device `inf * 0` as a non-finite value without asserting which one appears in silicon. The defect
is that the gradient is non-finite where the correct value is finite.

## Root cause

Both branches multiply by `reciprocal(input)`:

- `dim=None`:
  [`unary_backward.cpp#L1668-L1676`](https://github.com/tenstorrent/tt-metal/blob/a5cd86212e/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L1668-L1676)
  — `multiply(reciprocal(input), fill_first_val_into_tensor(prod * grad))`.
- per-dimension: `reciprocal_input` is built once at
  [`L1711`](https://github.com/tenstorrent/tt-metal/blob/a5cd86212e/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L1711)
  and feeds every `dim` branch, e.g. `dim=-1` at
  [`L1722`](https://github.com/tenstorrent/tt-metal/blob/a5cd86212e/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L1722).
  All four dimensions are affected, not only `dim=-1`.

At a zero input the product of the other values has already been lost in the reduction, so it
cannot be recovered by the division. The formula is valid only when the reduced input has no zeros.

## Reproducer

Host only, no device required (`pip install torch`):

```python
"""Source-level replay of ttnn.prod_bw against torch.prod autograd.

Replays the composite expression from tt-metal main @ a5cd86212e. Runs on host only:
it does not need a Tenstorrent device.
"""

import math
import torch


def source_prod_bw(values, grad):
    """unary_backward.cpp L1668-1676: reciprocal(input) * broadcast(prod(input) * grad)."""
    product = math.prod(values)
    out = []
    for v in values:
        if v == 0.0:
            # reciprocal(0) is +inf on device; inf * 0 is non-finite
            out.append(float("inf") * (product * grad))
        else:
            out.append(product * grad / v)
    return out


def source_prod_bw_dim(rows, grad):
    """unary_backward.cpp L1711-1722, dim=-1: bcast(reciprocal(input), prod * grad, MUL)."""
    return [source_prod_bw(row, grad) for row in rows]


def torch_prod_bw(values, grad):
    t = torch.tensor(values, dtype=torch.float64, requires_grad=True)
    torch.prod(t).backward(torch.tensor(grad, dtype=torch.float64))
    return t.grad.tolist()


def torch_prod_bw_dim(rows, grad):
    t = torch.tensor(rows, dtype=torch.float64, requires_grad=True)
    torch.prod(t, dim=-1, keepdim=True).backward(
        torch.full((len(rows), 1), grad, dtype=torch.float64))
    return t.grad.tolist()


def show(v):
    return "non-finite" if isinstance(v, float) and not math.isfinite(v) else f"{v:g}"


print("dim=None")
for values in ([2.0, 3.0, 4.0], [2.0, 0.0, 4.0], [0.0, 0.0, 4.0], [1.0, 5.0, 0.0]):
    actual, expected = source_prod_bw(values, 1.0), torch_prod_bw(values, 1.0)
    ok = all(a == e for a, e in zip(actual, expected))
    print(f"  input={values}  expected={expected}  "
          f"current=[{', '.join(show(a) for a in actual)}]  {'ok' if ok else 'DIVERGES'}")

print("dim=-1")
for rows in ([[2.0, 3.0, 4.0], [1.0, 5.0, 3.0]], [[2.0, 0.0, 4.0], [1.0, 5.0, 3.0]]):
    actual, expected = source_prod_bw_dim(rows, 1.0), torch_prod_bw_dim(rows, 1.0)
    ok = all(a == e for ar, er in zip(actual, expected) for a, e in zip(ar, er))
    print(f"  input={rows}  expected={expected}  "
          f"current=[{', '.join('[' + ', '.join(show(a) for a in r) + ']' for r in actual)}]  "
          f"{'ok' if ok else 'DIVERGES'}")
```

Output:

```text
dim=None
  input=[2.0, 3.0, 4.0]  expected=[12.0, 8.0, 6.0]  current=[12, 8, 6]  ok
  input=[2.0, 0.0, 4.0]  expected=[0.0, 8.0, 0.0]  current=[0, non-finite, 0]  DIVERGES
  input=[0.0, 0.0, 4.0]  expected=[0.0, 0.0, 0.0]  current=[non-finite, non-finite, 0]  DIVERGES
  input=[1.0, 5.0, 0.0]  expected=[0.0, 0.0, 5.0]  current=[0, 0, non-finite]  DIVERGES
dim=-1
  input=[[2.0, 3.0, 4.0], [1.0, 5.0, 3.0]]  expected=[[12.0, 8.0, 6.0], [15.0, 3.0, 5.0]]  current=[[12, 8, 6], [15, 3, 5]]  ok
  input=[[2.0, 0.0, 4.0], [1.0, 5.0, 3.0]]  expected=[[0.0, 8.0, 0.0], [15.0, 3.0, 5.0]]  current=[[0, non-finite, 0], [15, 3, 5]]  DIVERGES
```

## Expected behaviour

The gradient of a product with respect to one factor is the product of the *other* factors, which
is always finite for finite inputs:

- no zero in the reduced input: the current formula is correct;
- exactly one zero: zero everywhere except at that position, which gets `grad` times the product
  of the non-zero values;
- two or more zeros: zero everywhere.

Suggested tests: exact zeros in both paths, one zero and several zeros, and a non-finite check
before PCC — a PCC comparison alone can pass or silently degrade on a non-finite element.

### PR DESCRIPTION
**Pull request by 0xalydev**
### Summary

Resolves #54551.

### Root Cause
`ttnn.prod_bw` previously computed the gradient via `reciprocal(input) * (prod * grad)`. When an input element contains an exact zero, `reciprocal(0.0)` evaluates to `+inf`, and multiplying by `prod(input) == 0.0` produced non-finite (`inf * 0.0 -> NaN`) values on device instead of the mathematically correct finite gradient.

### Fix Details
1. **Zero-Aware Masking**:
   - Computes `is_zero = ttnn::eq(input, 0.0f)` and counts total zeros across reduction dimensions.
   - Replaces zero inputs with `1.0f` for safe reciprocal and non-zero product computations.
2. **Branching Gradient Logic**:
   - **Case 0 (No zeros)**: Standard fast-path `(prod * grad) / input`.
   - **Case 1 (Single zero)**: Position of the zero element evaluates to `prod(non_zero_elements) * grad`, and zero everywhere else.
   - **Case 2 (Two or more zeros)**: Gradient is finite `0.0f` everywhere.
3. Preserves layout and memory configurations across device tensors.

### CODE REVIEW

**Comment by jberkowitzTT** [comment]
Hi @0xalydev - someone else is already assigned to the associated bounty so I'll close this out in the meantime. If you're interested in the bounty please comment on the main issue. If you would like to be assigned a bounty before that one's reopened (we give assignees around 2 months to a bounty), please email us at bounties@tenstorrent.com & we'll look to review as soon as we can
